### PR TITLE
Add new Waffle tutorial

### PR DESCRIPTION
**Pull request by VladStarostenko**
## Description

Add new tutorial for Waffle about using matchers.

### CODE REVIEW

**Comment by samajammin** [comment]
Thanks for the submission @VladStarostenko! I've converted this PR to a "draft". Please let me know when it's complete & ready for review. 

I see there's a bunch of `.idea/` files in the PR as well - could you please remove those / add that directory to our .gitignore file in this PR?

Cheers!

**Comment by VladStarostenko** [comment]
> Thanks for the submission @VladStarostenko! I've converted this PR to a "draft". Please let me know when it's complete & ready for review.
> 
> I see there's a bunch of `.idea/` files in the PR as well - could you please remove those / add that directory to our .gitignore file in this PR?
> 
> Cheers!

Hi @samajammin ! 

This tutorial is ready for review 😃 

Cheers!

File: src/content/developers/tutorials/testing-smart-contract-with-waffle/index.md [hunk 1]
```diff
@@ -0,0 +1,1140 @@
+---
+title: Testing smart contract with Waffle
```
Comment (samajammin) [inline src/content/developers/tutorials/testing-smart-contract-with-waffle/index.md:2]:
```suggestion
title: Testing smart contracts with Waffle
```

File: src/content/developers/tutorials/testing-smart-contract-with-waffle/index.md [hunk 2]
```diff
@@ -0,0 +1,1140 @@
...
+More details you can read in [waffle documentation](https://ethereum-waffle.readthedocs.io/en/latest/#waffle-documentation) !
+
+## The quick tutorial {#the-quick-tutorial}
+
+First things first, create new `JavaScript` or `TypeScript` project ( I'll use `TS`, but if you use `JS` it's not a problem ) :
+
+Somewhat like this :
+
+<details>
+<summary>package.json</summary>
```
Comment (samajammin) [inline src/content/developers/tutorials/testing-smart-contract-with-waffle/index.md:42]: Didn't know about `<details>` & `<summary>` elements - neat!

File: src/content/developers/tutorials/testing-smart-contract-with-waffle/index.md [hunk 3]
```diff
@@ -0,0 +1,1140 @@
+---
+title: Testing smart contract with Waffle
+description: Tutorial with examples of using smart contract matchers with waffle
```
Comment (samajammin) [inline src/content/developers/tutorials/testing-smart-contract-with-waffle/index.md:3]:
```suggestion
description: Learn how test Solidity smart contracts and use smart contract matchers with Waffle.
```

File: src/content/developers/tutorials/testing-smart-contract-with-waffle/index.md [hunk 4]
```diff
@@ -0,0 +1,1140 @@
...
+- you've written a few tests in `JavaScript`,
+- you’ve used `yarn` or `npm`, JavaScripts’s package installer.
+
+Again, if any of these are untrue, or you don’t plan to reproduce the code in this article, you can likely still follow along just fine.
+
+## A few words about Waffle {#a-few-words-about-waffle}
+
+[Waffle](https://getwaffle.io) is the most advanced library for writing and testing smart contracts.
+
+Works with ethers-js.
```
Comment (samajammin) [inline src/content/developers/tutorials/testing-smart-contract-with-waffle/index.md:31]:
```suggestion
Works with the [JavaScript API](/developers/docs/apis/javascript/) ethers-js.
```

File: src/content/developers/tutorials/testing-smart-contract-with-waffle/index.md [hunk 5]
```diff
@@ -0,0 +1,1140 @@
...
+
+More details you can read in [waffle documentation](https://ethereum-waffle.readthedocs.io/en/latest/#waffle-documentation) !
```
Comment (samajammin) [inline src/content/developers/tutorials/testing-smart-contract-with-waffle/index.md:33]:
```suggestion
You can read more details in the [Waffle documentation](https://ethereum-waffle.readthedocs.io/en/latest/#waffle-documentation) !
```

File: src/content/developers/tutorials/testing-smart-contract-with-waffle/index.md [hunk 6]
```diff
@@ -0,0 +1,1140 @@
...
+    ]
+    }
+
+</details>
+
+## Step #1: Install waffle in your project [Link to doc](https://ethereum-waffle.readthedocs.io/en/latest/getting-started.html#installation) {#step-1-install-waffle-in-your-project}
+
+To get started, install `ethereum-waffle`. In this tutorial, I'll use `yarn`, so to install `ethereum-waffle` run:
+
+```
```
Comment (samajammin) [inline src/content/developers/tutorials/testing-smart-contract-with-waffle/index.md:644]:
```suggestion
```bash
```

File: src/content/developers/tutorials/testing-smart-contract-with-waffle/index.md [hunk 7]
```diff
@@ -0,0 +1,1140 @@
...
+ yarn add --dev ethereum-waffle
+```
+
+## Step #2: Write a smart contract [Link to doc](https://ethereum-waffle.readthedocs.io/en/latest/getting-started.html#writing-a-contract) {#step-2-write-a-smart-contract}
+
+In this tutorial, I'll use [ERC20](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/ded2b0a55c9c13731963ab7b85a70c8e73504bab/contracts/token/ERC20/ERC20.sol) token from [OpenZeppelin](https://openzeppelin.com).
+
+So, add `OpenZeppelin` by installing it with `yarn`:
+
+```
```
Comment (samajammin) [inline src/content/developers/tutorials/testing-smart-contract-with-waffle/index.md:654]:
```suggestion
```bash
```

File: src/content/developers/tutorials/testing-smart-contract-with-waffle/index.md [hunk 8]
```diff
@@ -0,0 +1,1140 @@
...
+```json
+{
+  "compilerType": "solcjs",
+  "compilerVersion": "0.6.2",
+  "sourceDirectory": "./src",
+  "outputDirectory": "./build"
+}
+```
+
+You can read more about the waffle configuration [here](https://ethereum-waffle.readthedocs.io/en/latest/configuration.html#configuration).
```
Comment (samajammin) [inline src/content/developers/tutorials/testing-smart-contract-with-waffle/index.md:699]:
```suggestion
You can read more about the Waffle configuration [here](https://ethereum-waffle.readthedocs.io/en/latest/configuration.html#configuration).
```

File: src/content/developers/tutorials/testing-smart-contract-with-waffle/index.md [hunk 9]
```diff
@@ -0,0 +1,1140 @@
...
+
+### Step #4.1 Install necessary dependencies [Link to doc](https://ethereum-waffle.readthedocs.io/en/latest/getting-started.html#writing-tests)
+
+After we have successfully authored a Smart Contract we can test it. We will use `Waffle` to do it.
+
+Tests in `Waffle` are written using `Mocha` alongside with `Chai`. We can use a different test environment, but `Waffle` matchers only work with `Chai`.
+
+So, we need to add `Chai` to our dependencies :
+
+```
```
Comment (samajammin) [inline src/content/developers/tutorials/testing-smart-contract-with-waffle/index.md:1021]:
```suggestion
```bash
```

**Review by samajammin — APPROVED** [review]
Looks great! Made a few tweaks, will commit those & merge. Thanks for adding this!

**Review by samajammin — APPROVED** [review]

**Pull request merged**